File: problems/91_decode-ways/main.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution
{
public:
    int numDecodings(string s)
    {
        int n = s.size();
        vector<int> dp(n + 1, -1);
        dp[n] = 1;
        for (int i = n - 1; i >= 0; i--)
        {
            if (s[i] == '0')
            {
                dp[i] = 0;
                continue;
            }
            dp[i] = dp[i + 1];
            if (i + 1 < n)
            {
                int val = (s[i] - '0') * 10 + (s[i + 1] - '0');
                if (val >= 10 && val <= 26)
                    dp[i] += dp[i + 2];
            }
        }
        return dp[0];
    }
};

int main()
{
    Solution solution;

    vector<pair<string, int>> samples = {
        {"12", 2},
        {"226", 3},
        {"06", 0},
    };

    for (size_t i = 0; i < samples.size(); ++i)
    {
        const string &s = samples[i].first;
        int expected = samples[i].second;
        int result = solution.numDecodings(s);

        cout << "Example " << (i + 1) << " result: " << result
             << ", expected: " << expected << '\n';
    }

    return 0;
}

File: problems/91_decode-ways/main.rs
struct Solution;

impl Solution {
    pub fn num_decodings(s: String) -> i32 {
        if s.is_empty() {
            return 0;
        }
        let bytes = s.as_bytes();
        if bytes[0] == b'0' {
            return 0;
        }

        let mut prev: i32 = 1; // dp[i-2]
        let mut curr: i32 = 1; // dp[i-1]

        for i in 1..bytes.len() {
            let mut next: i32 = 0;

            // single-digit decode (non-zero)
            if bytes[i] != b'0' {
                next += curr;
            }

            // two-digit decode (10..26)
            if bytes[i - 1] == b'1' || (bytes[i - 1] == b'2' && bytes[i] <= b'6') {
                next += prev;
            }

            prev = curr;
            curr = next;
        }

        curr
    }
}

fn main() {
    let samples: Vec<(&str, i32)> = vec![("12", 2), ("226", 3), ("06", 0)];

    for (idx, (s, expected)) in samples.iter().enumerate() {
        let result = Solution::num_decodings((*s).to_string());
        println!(
            "Example {} result: {}, expected: {}",
            idx + 1,
            result,
            expected
        );
    }
}
